Spigot-API-Patches/0136-Persistent-and-Temporary-Meta-API.patch

```diff
@@ -0,0 +1,1461 @@
+
+Temporary Metadata is a replacement API for Bukkit's Metadatable interface,
+that provides an API that doesn't suck, and is simpler (ultimately just a hashmap)
+
+diff --git a/src/main/java/com/destroystokyo/paper/meta/Meta.java b/src/main/java/com/destroystokyo/paper/meta/Meta.java
+new file mode 100644
+index 000000000..bfe6e3242
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/meta/Meta.java
+@@ -0,0 +1,43 @@
++package com.destroystokyo.paper.meta;
++
++import org.bukkit.NamespacedKey;
++import org.bukkit.inventory.ItemStack;
++import org.bukkit.plugin.Plugin;
++
++@SuppressWarnings("WeakerAccess")
++public class Meta {
++
++    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
++    public static boolean isValidPersistentMeta(Object value) {
++        return (value instanceof String) || (value instanceof Long) ||
++            (value instanceof Integer) || (value instanceof ItemStack) ||
++            (value instanceof Float) || (value instanceof Double) ||
++            (value instanceof PersistentMetaMap) || (value instanceof PersistentMetaList) ||
++            (value instanceof PersistentMetaSet);
++        // TODO: ConfigurationSerializable
++    }
++
++    public static TempMetaKey createTempKey(String key) {
++        return new TempMetaKey(null, key);
++    }
++
++    public static PersistentMetaKey createPersistentKey(String key) {
++        return new PersistentMetaKey(null, key);
++    }
++
++    public static TempMetaKey createTempKey(Plugin plugin, String key) {
++        return new TempMetaKey(plugin.getName(), key);
++    }
++
++    public static PersistentMetaKey createPersistentKey(Plugin plugin, String key) {
++        return new PersistentMetaKey(plugin.getName(), key);
++    }
++
++    public static TempMetaKey createTempKey(NamespacedKey key) {
++        return new TempMetaKey(key.getNamespace(), key.getKey());
++    }
++
++    public static PersistentMetaKey createPersistentKey(NamespacedKey key) {
++        return new PersistentMetaKey(key.getNamespace(), key.getKey());
++    }
++}
+diff --git a/src/main/java/com/destroystokyo/paper/meta/MetaContainer.java b/src/main/java/com/destroystokyo/paper/meta/MetaContainer.java
+new file mode 100644
+index 000000000..a03f915d3
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/meta/MetaContainer.java
+@@ -0,0 +1,408 @@
++package com.destroystokyo.paper.meta;
++
++import org.bukkit.inventory.ItemStack;
++
++import java.util.List;
++import java.util.Map;
++import java.util.Set;
++
++/**
++ * Represents an object that can contain metadata.
++ * <p>
++ * This is abstract from the Persistent vs temporary context
++ * <p>
++ * This API is not thread safe!
++ */
++@SuppressWarnings({"unchecked", "unused"})
++public interface MetaContainer {
++
++    MetaMap getMetaMap(MetaKey key, boolean isWrite);
++
++    TempMetaMap getTempMetaMap(TempMetaKey key, boolean isWrite);
++
++    default <T> T getMeta(TempMetaKey key) {
```

Comment:
Annotate the method with `@Nullable` - applies to all methods that can return `null`.